test/end2end_test.go

```diff
@@ -193,6 +193,16 @@ func (s *testServer) HalfDuplexCall(stream testpb.TestService_HalfDuplexCallServ
 
 const tlsDir = "testdata/"
 
+func TestDialTimeout(t *testing.T) {
+	conn, err := grpc.Dial("localhost:0", grpc.WithTimeout(5*time.Microsecond))
```

Comment:
Hrm, I'm not sure I need to start up a server here, since the easiest way to test this timeout is to attempt connecting to an unopened socket.

So, how to go about getting a socket we're sure is unopened--in the flaky case, perhaps some other test leaves one open on accident? I've implemented a simplistic open/close/test-with-recently-closed-address logic, but I'm not really happy with it. If you have a better idea, I'm all ears :) but it's 2AM here and I'll have to take another look in the morning.